### PR TITLE
Enable random rules to be populated from the wiki guidelines

### PR DESCRIPTION
**Pull request by cpeel**
The random rules shown on the round pages are pulled from the `rules` table in the database. This table was originally populated by `SETUP/rebuild_random_rules.sh` from the English documentation in the code. This also meant that Random Rules were always only presented in English, even if we had rules in the user's language.

When pgdp.net moved to having the documentation in the Wiki, we lost the ability to update these rules from the authoritative source. This is going to be most notable when we move to Unicode and need to update our documentation.

This MR centralizes all access to the `rules` table inside a single RandomRule class. It allows rules to be populated from a URL, which would be a pgdp.net-based wiki URL or -- for other sites -- the documentation that still exists within the wiki. Rules are imported for a specific language and shown to the user in their language if available.

This branch is testable in my [random-rules](https://www.pgdp.org/~cpeel/c.branch/random-rules) sandbox, specifically see the [Manage Random Rules](https://www.pgdp.org/~cpeel/c.branch/random-rules/tools/site_admin/manage_random_rules.php) SA tool.

### CODE REVIEW

File: tools/site_admin/manage_random_rules.php [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+$url = array_get($_POST, "url", NULL);
+$langcode = strtolower(array_get($_POST, "langcode", NULL));
+
+$title = _("Manage Random Rules");
+
+output_header($title, NO_STATSBAR);
+
+echo "<h1>" . html_safe($title) . "</h1>";
+
+echo "<p>" . _("On the round pages, users are presented a random rule from the proofreading or formatting guidelines. These rules are stored in the database. The database is populated from the guidelines directly, either the guidelines contained within the code or those in the pgdp.net wiki (or similarly-formatted rules), via URL. If the guidelines are updated, the rules will need to be reloaded using this page. Rules are managed as a language-document unit and are deleted or updated as that unit.") . "</p>";
```
Comment (srjfoo) [inline tools/site_admin/manage_random_rules.php]: Sometimes it helps to be explicit ...

> Rules are managed as a language-document unit and are deleted or updated as that unit`, and cannot be replaced or add as individual rules.`

File: pinc/RandomRule.inc [hunk 2]
```diff
@@ -0,0 +1,266 @@
+<?php
+include_once($relPath."misc.inc"); // startswith()
+
+class RandomRule
+{
+    private $table_row = [];
+    public static $document_values = [
+        "proofreading_guidelines.php" => "Proofreading Guidelines",
+        "formatting_guidelines.php" => "Formatting Guidelines",
```
[the remark this answers is not in the document]
Comment (cpeel) [reply pinc/RandomRule.inc]: It does. These strings are only used in the admin interface. As static variables they can't be declared with `_()`. I can easily move them within a function and translate them if you think that would be better.

File: pinc/RandomRule.inc [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        return isset($this->table_row[$name]);
+    }
+
+    public function load($id)
+    {
+        $sql = sprintf("
+            SELECT *
+            FROM rules
+            WHERE id = %d
+        ", $id);
```
[the remark this answers is not in the document]
Comment (cpeel) [reply pinc/RandomRule.inc]: It is `id` in the rules table. It's an auto-generated primary key that has no value outside of being able to uniquely identify a given rule.

File: tools/site_admin/manage_random_rules.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+echo "<h2>" . _("Update random rules") . "</h2>";
+
+echo "<p>" . _("To update rules -- either adding rules for a new language or replacing rules for an existing langauge -- specify a two-letter language code, the document type, and a full URL to parse. The URL can be a pgdp.net wiki page (or a similarly formatted page) or documentation included with the code.") . "</p>";
+
+echo "<p>" . _("To preview what loaded rules will look like, you can use a nonexistent language code such as 'test', view the rules, and then load into the correct language.") . "</p>";
+
+echo "<form method='POST'>";
+echo "<input type='hidden' name='action' value='update'>";
+echo _("Language") . ": <input type='text' name='langcode' size='4' required> ";
```
[the remark this answers is not in the document]
Comment (cpeel) [reply tools/site_admin/manage_random_rules.php]: I'll add a `maxsize` to that field to limit input to 4 characters (although the DB column should have been 5, so if it was 4 I did it wrong anyway).

File: pinc/RandomRule.inc (hunk 2 above)
[the remark this answers is not in the document]
Comment (srjfoo) [reply pinc/RandomRule.inc]: If they're only in the admin interface, let's not bother.

File: pinc/RandomRule.inc (hunk 3 above)
[the remark this answers is not in the document]
Comment (srjfoo) [reply pinc/RandomRule.inc]: Ah, okay, thanks. I was *totally* misunderstanding what I was reading, then.

File: tools/site_admin/manage_random_rules.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (srjfoo) [reply tools/site_admin/manage_random_rules.php]: Yes, the DB column is 5 -- I just figured 4 was hedging a bit.

File: tools/site_admin/manage_random_rules.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+echo "</table>";
+
+//----------------------------------------------------------------------------
+
+echo "<h2>" . _("Update random rules") . "</h2>";
+
+echo "<p>" . _("To update rules -- either adding rules for a new language or replacing rules for an existing langauge -- specify a two-letter language code, the document type, and a full URL to parse. The URL can be a pgdp.net wiki page (or a similarly formatted page) or documentation included with the code.") . "</p>";
+
+echo "<p>" . _("To preview what loaded rules will look like, you can use a nonexistent language code such as 'test', view the rules, and then load into the correct language.") . "</p>";
```
[the remark this answers is not in the document]
Comment (cpeel) [reply tools/site_admin/manage_random_rules.php]: Updated, thanks!

File: tools/site_admin/manage_random_rules.php (hunk 1 above)
Comment (cpeel) [reply tools/site_admin/manage_random_rules.php]: Good call - updated.

File: pinc/RandomRule.inc (hunk 2 above)
[the remark this answers is not in the document]
Comment (cpeel) [reply pinc/RandomRule.inc]: I at least documented why we can't and what to do if we want to in the future.

**Comment by srjfoo** [comment]
After debugging why the `subject` field wasn't being populated from the Dutch guidelines (the code doesn't like wiki markup in that field), I think we need some instructions detailing how a rule should be marked (whether in the code proper, or in the wiki) so that it can be properly parsed. Casey suggested a markdown file that can be referenced both in comments in the code and from the Random Rule admin interface. This should be documented before merge.

File: pinc/RandomRule.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+The rules are populated directly from the rendered HTML versions of the
+guidelines via the Manage Random Rules page off the Site Admin UI. There
+are two versions of HTML supported:
+
+* HTML from the `faq/proofreading_guidelines.php` and `faq/formatting-guidelines.php`
+  pages within the code.
+* HTML from the [Proofreading Guidelines](https://www.pgdp.net/wiki/DP_Official_Documentation:Proofreading/Proofreading_Guidelines)
+  and [Formatting Guidelines](https://www.pgdp.net/wiki/DP_Official_Documentation:Formatting/Formatting_Guidelines)
+  from the pgdp.net Official Documentation.
```
Comment (srjfoo) [inline pinc/RandomRule.md]: Add "in the DP wiki" after "from the pgdp.net Official Documentation"?

File: pinc/RandomRule.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+Rules in the code-based guidelines are assumed to be in the following format:
+
+```html
+<h3><a name="ANCHOR">SUBJECT</a></h3>
+RULE (multiple lines) ...
+<!-- END RR -->
+```
+
+The end of the rule is either the start of another rule, or the appearance of
+`<!-- END RR -->`, whichever comes first.
```
[the remark this answers is not in the document]
Comment (cpeel) [reply pinc/RandomRule.md]: While you can add HTML comments to wiki page they are not included in the final HTML when the page is rendered, so they aren't there to find them -- hence why we had to find another solution for the wiki pages.

Honestly, I'd rather not commit to maintaining the in-code versions. That is one of the sacrifices we made, whether intentional or not, when we moved the DP docs into the wiki. There are probably numerous ways where the now-in-the-wiki docs are not HTML5 compliant.

[the remark this answers is not in the document]
Comment (srjfoo) [reply pinc/RandomRule.md]: I totally understand -- and I'm not really in favor of trying to maintain the in-code versions, either. I just feel guilty about not being explicit about anything we leave in the code not being up-to-date without it being explicitly stated that it isn't.

**Pull request merged**